salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,430 @@
+
+
+# }}}
+# Helpers {{{
+
+
+def _get_volume(name):
+    """Get a Volume object from the pillar."""
+    volume = __pillar__['metalk8s']['volumes'].get(name)
+    if volume is None:
+        raise ValueError('volume {} not found in pillar'.format(name))
+    if 'rawBlockDevice' in volume['spec']:
+        return RawBlockDevice(volume)
+    elif 'sparseLoopDevice' in volume['spec']:
+        return SparseLoopDevice(volume)
+    else:
+        raise ValueError('unsupported Volume type')
+
+
+def _run_cmd(cmd):
+    """Execute the given `cmd` command and return its result.
+
+    Raise `CommandExecutionError` if the command failed.
+
+    Args:
+        cmd  (str): command to execute
+
+    Returns:
+        dict: the command result (stderr, stdout, retcode, …)
+    """
+    ret = __salt__['cmd.run_all'](cmd)
+    if ret.get('retcode', 0) != 0:
+        raise CommandExecutionError(
+            'error while trying to run `{0}`: {1}' .format(cmd, ret['stderr'])
+        )
+    return ret
+
+
+def _quantity_to_bytes(quantity):
+    """Return a quantity with a unit converted into a number of bytes.
+
+    Examples:
+    >>> quantity_to_bytes('42Gi')
+    45097156608
+    >>> quantity_to_bytes('100M')
+    100000000
+    >>> quantity_to_bytes('1024')
+    1024
+
+    Args:
+        quantity (str): a quantity, composed of a count and an optional unit
+
+    Returns:
+        int: the capacity (in bytes)
+    """
+    UNIT_FACTOR = {
+      None:  1,
+      'Ki':  2 ** 10,
+      'Mi':  2 ** 20,
+      'Gi':  2 ** 30,
+      'Ti':  2 ** 40,
+      'Pi':  2 ** 50,
+      'k':  10 ** 3,
+      'M':  10 ** 6,
+      'G':  10 ** 9,
+      'T':  10 ** 12,
+      'P':  10 ** 15,
+    }
+    size_regex = r'^(?P<size>[1-9][0-9]*)(?P<unit>[kKMGTP]i?)?$'
+    match = re.match(size_regex, quantity)
+    assert match is not None, 'invalid resource.Quantity value'
+    size = int(match.groupdict()['size'])
+    unit = match.groupdict().get('unit')
+    return size * UNIT_FACTOR[unit]
+
+
+@contextlib.contextmanager
+def _open_fd(*args, **kwargs):
+    fd = os.open(*args, **kwargs)
+    try:
+        yield fd
+    finally:
+        os.close(fd)
+
+
+# Cannot use `disk.fstype` from Salt because you can't trust its output…
+#
+# Example:
+# - On a unformated sparse file:
+#     __salt__['disk.fstype']('/var/lib/metalk8s/storage/sparse/example')
+#     returns "/dev/sda1      ext4  41152736 7765776  31273476  20% /"
+#
+# - On a raw (non-formatted) block device:
+#     __salt__['disk.fstype']('/dev/sdb')
+#     returns "devtmpfs       devtmpfs   1932084     0   1932084   0% /dev"
+#
+# So yeah, let's not rely on this…
+def _get_fstype(path):
+    try:
+        res = _run_cmd('blkid {}'.format(path))
+        pat = r'TYPE="(?P<fstype>.+)"'
+        match = re.search(pat, res['stdout'])
+        return match.groupdict().get('fstype') if match else None
+    except CommandExecutionError:  # Happens on non-formatted sparse file.
```

Comment:
Could we then assert the given path is a file?

Reply:
What do you mean?

I'll rewrite this part to be more fine-grained (return code==2 I'll treat as "No FS", return code non-zero and !=2 I'll keep raising).

Reply:
OK, works for me.

I just wanted to make sure that `CommandExecutionError` (i.e., `rc != 0`) doesn't implicitly mean "no FS present", while there are plenty of other reasons why `blkid` can fail...

Reply:
Looking at the man page, checking for rc 2 catches quite a lot of things as well, including "impossible to gather any information"...

Reply:
Yeah you're right.

"impossible to gather any information" is a bit like `lsblk` failing on non-block device: in the end you don't know if it's formatted or not :/

I guess we could use `file` but the output format is a mess…

```
# truncate -s 1G bar
# file -sbL bar
data
# truncate -s 2G foo
# mkfs.xfs -f foo > /dev/null
# file -sbL foo
SGI XFS filesystem data (blksz 4096, inosz 512, v2 dirs)
# wipefs -af /dev/sdb > /dev/null
# file -sbL /dev/sdb
data
# mkfs.ext4 -F /dev/sdb > /dev/null
mke2fs 1.42.9 (28-Dec-2013)
# file -sbL /dev/sdb
Linux rev 1.0 ext4 filesystem data, UUID=e3276629-5051-431b-83b5-da808679f77a (extents) (64bit) (large files) (huge files)
```

Reply:
Done